File: company/company/report/purchase_report/purchase_report.js
// Copyright (c) 2025, deepak and contributors
// For license information, please see license.txt

frappe.query_reports["Purchase Report"] = {
    auto_run: true,

    filters: [
        { fieldname: "page", fieldtype: "Int", default: 1, hidden: 1 },
        { fieldname: "page_length", fieldtype: "Int", default: 10, hidden: 1 },

        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.add_months(frappe.datetime.get_today(), -12)
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.get_today()
        },
        {
            fieldname: "purchase_id",
            label: __("Purchase ID"),
            fieldtype: "Link",
            options: "Purchase"
        },
        {
            fieldname: "vendor_id",
            label: __("Vendor ID"),
            fieldtype: "Link",
            options: "Customer",
            get_query: () => {
                return {
                    filters: {
                        customer_type: "Purchase"
                    }
                };
            }
        },
        // {
        //     fieldname: "vendor_name",
        //     label: __("Vendor Name"),
        //     fieldtype: "Data"
        // },
        {
            fieldname: "gst_non_gst",
            label: __("GST / Non-GST"),
            fieldtype: "Select",
            options: "\nGST\nNon-GST"
        },
        {
            fieldname: "purchase_status",
            label: __("Purchase Status"),
            fieldtype: "Select",
            options: "\nPending\nPartially Paid\nFully Paid"
        },
        {
            fieldname: "business_person_name",
            label: __("Business Person"),
            fieldtype: "Link",
            options: "Business Person"
        }
    ],

    // ----------------------------
    // SETUP (RUNS ONCE)
    // ----------------------------
    onload(report) {
        report.set_filter_value("page_length", 10);

        report.page.add_inner_button(__("Export Purchase Details"), function () {
            let filters = report.get_filter_values(true) || {};

            frappe.call({
                method: 'company.company.api.get_purchase_export_count',
                args: {
                    filters: JSON.stringify(filters)
                },
                callback: function (r) {
                    let counts = r.message || { purchase_count: 0, item_count: 0 };
                    let pur_count = counts.purchase_count || 0;

                    if (pur_count === 0) {
                        frappe.msgprint({
                            title: __('No Purchases Found'),
                            message: __('There are no purchases matching the selected filters to export.'),
                            indicator: 'orange'
                        });
                        return;
                    }

                    frappe.confirm(
                        __('Are you sure you want to download <b>{0} Purchase(s)</b> to Excel?', [pur_count]),
                        function () {
                            frappe.dom.freeze(__('Generating Excel Report, please wait...'));

                            fetch('/api/method/company.company.api.export_purchase_itemized_excel', {
                                method: 'POST',
                                headers: {
                                    'Content-Type': 'application/x-www-form-urlencoded',
                                    'X-Frappe-CSRF-Token': frappe.csrf_token
                                },
                                body: $.param({
                                    filters: JSON.stringify(filters)
                                })
                            })
                            .then(response => response.blob())
                            .then(blob => {
                                frappe.dom.unfreeze();
                                let url = window.URL.createObjectURL(blob);
                                let a = document.createElement('a');
                                a.href = url;
                                a.download = 'Purchase_Item_Details_Report.xlsx';
                                document.body.appendChild(a);
                                a.click();
                                a.remove();
                                window.URL.revokeObjectURL(url);
                                frappe.show_alert({ message: __('Excel Report downloaded successfully!'), indicator: 'green' });
                            })
                            .catch(err => {
                                frappe.dom.unfreeze();
                                frappe.msgprint(__('Failed to generate Excel report. Please try again.'));
                            });
                        }
                    );
                }
            });
        });

        // Store button references
        report._prev_btn = report.page.add_inner_button("⬅ Prev", () => {
            const page = report.get_filter_value("page");
            if (page > 1) {
                report.set_filter_value("page", page - 1);
                report.refresh();
            }
        });

        report._next_btn = report.page.add_inner_button("Next ➡", () => {
            const page = report.get_filter_value("page");
            report.set_filter_value("page", page + 1);
            report.refresh();
        });

        // Ensure "All" data is exported even when view is paginated and filters are hidden
        report.export_report = () => {
            const dialog = frappe.report_utils.get_export_dialog(
                __(report.report_name),
                [],
                ({ file_format }) => {
                    const filters = report.get_filter_values(true);
                    // Force full data for export
                    filters.page_length = 999999;
                    filters.page = 1;
                    filters.is_export = 1;

                    const args = {
                        cmd: "frappe.desk.query_report.export_query",
                        report_name: report.report_name,
                        file_format_type: file_format,
                        filters: filters,
                        visible_idx: [], // Clear this to ensure server sends everything
                        is_export: 1,    // Signal to backend
                        include_indentation: 0,
                        include_filters: 0,
                        export_in_background: 0
                    };

                    open_url_post(frappe.request.url, args);
                }
            );
            dialog.show();
        };
    },

    // ----------------------------
    // DATA LOGIC
    // ----------------------------
    after_refresh(report) {
        const page = report.get_filter_value("page");
        const page_length = report.get_filter_value("page_length");

        setTimeout(() => {
            const datatable = report.datatable;
            if (!datatable || !datatable.datamanager) return;

            const rows = datatable.datamanager.getRows();

            // Disable Prev on first page
            if (report._prev_btn) {
                report._prev_btn.prop("disabled", page <= 1);
            }

            // Last page detection
            const is_last_page = rows.length < page_length;

            // Disable / enable Next
            if (report._next_btn) {
                report._next_btn.prop("disabled", is_last_page);
            }

            // Safety: if user goes beyond last page
            if (rows.length === 0 && page > 1) {
                report.set_filter_value("page", page - 1);
                report.refresh();
            }

        }, 0);
    }
};

File: company/company/report/purchase_report/purchase_report.py
import frappe
import json


@frappe.whitelist()
def execute(filters=None):

    # ---------------------------------------------
    # Parse filters
    # ---------------------------------------------
    if isinstance(filters, str):
        filters = json.loads(filters)

    filters = filters or {}

    # -------------------------------
    # PAGINATION
    # -------------------------------
    page = int(filters.get("page", 1))
    page_length = int(filters.get("page_length", 10))
    offset = (page - 1) * page_length

    # ---------------------------------------------
    # COLUMNS
    # ---------------------------------------------
    columns = [
        {"fieldname": "row_no", "label": "#", "fieldtype": "Int", "width": 60},
        {
            "fieldname": "purchase_no",
            "label": "Purchase ID",
            "fieldtype": "Link",
            "options": "Purchase",
            "width": 150,
        },
        {
            "fieldname": "bill_date",
            "label": "Bill Date",
            "fieldtype": "Date",
            "width": 120,
        },
        {
            "fieldname": "vendor_id",
            "label": "Vendor ID",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 140,
        },
        {
            "fieldname": "vendor_name",
            "label": "Vendor Name",
            "fieldtype": "Data",
            "width": 180,
        },
        {
            "fieldname": "bill_no",
            "label": "Bill No",
            "fieldtype": "Data",
            "width": 140,
        },
        {"fieldname": "qty", "label": "Qty", "fieldtype": "Float", "width": 80},
        {
            "fieldname": "grand_total",
            "label": "Purchase Total",
            "fieldtype": "Currency",
            "width": 140,
        },
        {
            "fieldname": "paid_amount",
            "label": "Paid Amount",
            "fieldtype": "Currency",
            "width": 130,
        },
        {
            "fieldname": "balance_amount",
            "label": "Balance Amount",
            "fieldtype": "Currency",
            "width": 130,
        },
        {
            "fieldname": "purchase_status",
            "label": "Status",
            "fieldtype": "Data",
            "width": 130,
        },
    ]

    # ---------------------------------------------
    # CONDITIONS
    # ---------------------------------------------
    conditions = []
    values = {}

    if filters.get("from_date"):
        conditions.append("p.bill_date >= %(from_date)s")
        values["from_date"] = filters["from_date"]

    if filters.get("to_date"):
        conditions.append("p.bill_date <= %(to_date)s")
        values["to_date"] = filters["to_date"]

    if filters.get("purchase_id"):
        conditions.append("p.name = %(purchase_id)s")
        values["purchase_id"] = filters["purchase_id"]

    if filters.get("vendor_id"):
        conditions.append("p.vendor_id = %(vendor_id)s")
        values["vendor_id"] = filters["vendor_id"]

    if filters.get("vendor_name"):
        conditions.append("p.vendor_name LIKE %(vendor_name)s")
        values["vendor_name"] = f"%{filters['vendor_name']}%"

    if filters.get("purchase_status"):
        conditions.append("p.purchase_status = %(purchase_status)s")
        values["purchase_status"] = filters["purchase_status"]
    
    # GST / Non-GST filter
    if filters.get("gst_non_gst"):
        if filters["gst_non_gst"] == "Non-GST":
            # Non-GST = Exempted tax
            conditions.append("p.default_tax_type = 'Exempted'")
        elif filters["gst_non_gst"] == "GST":
            # GST = any tax other than Exempted
            conditions.append("p.default_tax_type != 'Exempted'")

    where_clause = "WHERE " + " AND ".join(conditions) if conditions else ""

    # ---------------------------------------------
    # MAIN QUERY (WITH PAGINATION AND AGGREGATION)
    # ---------------------------------------------
    data = frappe.db.sql(
        f"""
        SELECT
            p.name              AS purchase_no,
            p.bill_date,
            p.vendor_id,
            p.vendor_name,
            p.bill_no,
            SUM(i.quantity)     AS qty,
            SUM(i.sub_total) / NULLIF(SUM(i.quantity), 0) AS price,
            p.grand_total,
            p.paid_amount,
            p.balance_amount,
            p.purchase_status
        FROM `tabPurchase` p
        LEFT JOIN `tabPurchase Items` i
            ON i.parent = p.name
        {where_clause}
        GROUP BY p.name
        ORDER BY p.bill_date DESC, p.creation DESC
        LIMIT %(limit)s OFFSET %(offset)s
        """,
        {**values, "limit": page_length, "offset": offset},
        as_dict=True,
    )

    # -------------------------------
    # ROW NUMBERS
    # -------------------------------
    for i, row in enumerate(data, start=offset + 1):
        row["row_no"] = i

    # ---------------------------------------------
    # TOTAL COUNT
    # ---------------------------------------------
    total_count = frappe.db.sql(
        f"SELECT COUNT(*) FROM `tabPurchase` p {where_clause}",
        values
    )[0][0]

    # ---------------------------------------------
    # SUMMARY (PURCHASE-LEVEL, FILTER AWARE)
    # ---------------------------------------------
    totals = frappe.db.sql(
        f"""
        SELECT
            SUM(p.grand_total) AS total_purchase,
            SUM(p.paid_amount) AS total_paid,
            SUM(p.balance_amount) AS total_pending
        FROM `tabPurchase` p
        {where_clause}
        """,
        values,
        as_dict=True,
    )[0]

    total_purchase = totals.total_purchase or 0
    total_paid = totals.total_paid or 0
    total_pending = totals.total_pending or 0

    paid_percent = (total_paid / total_purchase * 100) if total_purchase else 0
    pending_percent = (total_pending / total_purchase * 100) if total_purchase else 0

    report_summary = [
        {
            "label": "Total Purchase Amount",
            "value": total_purchase,
            "indicator": "blue",
            "datatype": "Currency",
        },
        {
            "label": f"Paid Amount ({paid_percent:.1f}%)",
            "value": total_paid,
            "indicator": "green" if paid_percent >= 50 else "orange",
            "datatype": "Currency",
        },
        {
            "label": f"Pending Amount ({pending_percent:.1f}%)",
            "value": total_pending,
            "indicator": "red" if total_pending > 0 else "green",
            "datatype": "Currency",
        },
    ]

    summary_data = {
        "purchase_count": total_count,
        "page": page,
        "page_length": page_length
    }

    return columns, data, summary_data, None, report_summary
